### DISCUSSION TITLE
Connecting with TestRail

### DISCUSSION THREAD

**Discussion by ss-napier** [seleniumbase/SeleniumBase#2712] [Q&A] [answered]
Hey there,

So Ideally I have a tool to connect seleniumbase with testrail, I managed to get it working (almost) by going through this guide
https://github.com/gurock/automation-frameworks-integration/tree/main/samples/python/pytest
https://github.com/gurock/automation-frameworks-integration/blob/main/samples/python/pytest-selenium/tests/conftest.py

So the thing is, I would like to able to modify the junitxml with the record_property so that it can ammend screenshot file to the xml reporting. 

Is there a way to modify the xml in seleniumbase without forking the whole project?

thanks in advance!

**Comment by mdmintz** [comment, marked as answer]
The `--junit-xml=report.xml` option comes from `pytest`, not `seleniumbase`.
See: https://github.com/pytest-dev/pytest/blob/5cffef7f07a1007681b671304e4ee69864c13940/src/_pytest/junitxml.py#L387

You can save images in base64 format so that you can more easily add it to the report after you modify the file.

You may want to use SeleniumBase's built-in report system:
See [Dashboard / Reports / Logs](https://github.com/seleniumbase/SeleniumBase/blob/master/examples/example_logs/ReadMe.md).

**Comment by ss-napier** [comment]
Okey based from your answer above, means I cant proceed to add a value in the xml report even with the selenium built in report yes?

**Reply by mdmintz** [reply]
You can always use Python file I/O to modify any files after they been created.

**Reply by ss-napier** [reply]
Thanks @mdmintz! I've created a temporary solution based on your suggestion. will share it below

**Comment by ss-napier** [comment]
For future reference, those who use testrail <> seleniumbase but dont want to fork/download the whole project, you can create a separate script and run it after your test complete.

Example of the code I used.

```
import xml.etree.ElementTree as ET
import os

def modify_junit_xml(path_to_junit_report, output_file):
    tree = ET.parse(path_to_junit_report)
    root = tree.getroot()

    testcases = root.findall(".//testcase")

    # Match attachments to test cases based on classname and name attributes
    attachments = {}
    for attachment_filename in os.listdir(attachment_folder):
        attachment_path = os.path.join(attachment_folder, attachment_filename)
        if os.path.isfile(attachment_path):
            attachments[attachment_filename] = attachment_path

    # Modify each testcase element
    for testcase in testcases:
        classname = testcase.get("classname")
        name = testcase.get("name")
        attachment_filepath = f"{classname}.{name}.png" if attachment_folder else ""
        if attachment_filepath in attachments:
             add_attachment_to_testcase(testcase, "testrail_attachment", attachments[attachment_filepath])
 
    tree.write(output_file)

def add_attachment_to_testcase(testcase_element, sub_element_name, sub_element_value):
    properties_element = testcase_element.find("properties")
    if properties_element is None:
        properties_element = ET.SubElement(testcase_element, "properties")
        
    property_element = ET.SubElement(properties_element, "property")
    property_element.set("name", sub_element_name)
    property_element.set("value", sub_element_value) #this can be path to your screenshot file path
```

Steps:
1. Make sure to run your test(s) with this --junit-xml="/report/junit-report.xml" 
2. Run above script once junit-report.xml is generated together with the screenshot 
3. then use the [trcli](https://support.testrail.com/hc/en-us/articles/7146548750868-Overview-and-installation#InstallingtheTestRailCLI) command and upload the modified xml file.

use below testrail xml mapping to get the sub_element_name
[TestRail Junit Reference](https://support.testrail.com/hc/en-us/articles/12989737200276-JUnit-to-TestRail-mapping#h_01H9QJ39FKSDG9B7WNYDCF49AE)


👍if you find this useful.
